# Section 3 Snowshoe 

# 1) Archive Exploration

- What is this data about?
- During what time frame were the observations in the dataset collected?
Does the dataset contain sensitive data?
Is there a publication associated with this dataset?




- i) This dataset references physical and demographical data on a "keystone" prey species in the northern boreal forest that experiences fluctuations in population from 8 - 11 years. The txt file contains observational variables such as: ear length, sex, age, weight, ear tag ID and type of survey.
- It is noted that this prey species population cycles is cornerstone to the influences of vegetation and predators.

- ii) This capture and re-capture took place for 2 - 3 years





# Citation

Kielland, K., F.S. Chapin, R.W. Ruess, and Bonanza Creek LTER. 2017. Snowshoe hare physical data in Bonanza Creek Experimental Forest: 1999-Present ver 22. Environmental Data Initiative.

# 2) Adding An Image

![Snowshoe Rabbit](https://upload.wikimedia.org/wikipedia/commons/thumb/8/8a/SNOWSHOE_HARE_%28Lepus_americanus%29_%285-28-2015%29_quoddy_head%2C_washington_co%2C_maine_-01_%2818988734889%29.jpg/1089px-SNOWSHOE_HARE_%28Lepus_americanus%29_%285-28-2015%29_quoddy_head%2C_washington_co%2C_maine_-01_%2818988734889%29.jpg?20170313021652)

# 3) Data Loading and Preliminary Exploration

In [10]:
#Importing libraries
import pandas as pd
import numpy as np

hares = pd.read_csv('data/55_Hare_Data_2012.txt')

In [13]:
hares.head()

date time    grid trap       l_ear r_ear  sex  age  weight  hindft  \
0  11/26/1998  NaN  bonrip   1A  414D096A08   NaN  NaN  NaN  1370.0   160.0   
1  11/26/1998  NaN  bonrip   2C  414D320671   NaN    M  NaN  1430.0     NaN   
2  11/26/1998  NaN  bonrip   2D  414D103E3A   NaN    M  NaN  1430.0     NaN   
3  11/26/1998  NaN  bonrip   2E  414D262D43   NaN  NaN  NaN  1490.0   135.0   
4  11/26/1998  NaN  bonrip   3B  414D2B4B58   NaN  NaN  NaN  1710.0   150.0   

  notes  b_key  session_id       study  
0   NaN  917.0          51  Population  
1   NaN  936.0          51  Population  
2   NaN  921.0          51  Population  
3   NaN  931.0          51  Population  
4   NaN  933.0          51  Population

In [14]:
print(hares.head(), end = "\n")
print(hares.tail(), end = "\n")
print(hares.info(), end = "\n")

         date time    grid trap       l_ear r_ear  sex  age  weight  hindft  \
0  11/26/1998  NaN  bonrip   1A  414D096A08   NaN  NaN  NaN  1370.0   160.0   
1  11/26/1998  NaN  bonrip   2C  414D320671   NaN    M  NaN  1430.0     NaN   
2  11/26/1998  NaN  bonrip   2D  414D103E3A   NaN    M  NaN  1430.0     NaN   
3  11/26/1998  NaN  bonrip   2E  414D262D43   NaN  NaN  NaN  1490.0   135.0   
4  11/26/1998  NaN  bonrip   3B  414D2B4B58   NaN  NaN  NaN  1710.0   150.0   

  notes  b_key  session_id       study  
0   NaN  917.0          51  Population  
1   NaN  936.0          51  Population  
2   NaN  921.0          51  Population  
3   NaN  931.0          51  Population  
4   NaN  933.0          51  Population  
          date      time    grid trap l_ear r_ear  sex  age  weight  hindft  \
3375  8/8/2002  18:00:00  bonrip  1b   1201  1202  NaN  NaN  1400.0     NaN   
3376  8/8/2002   6:00:00  bonrip  4b   1201  1202  NaN  NaN     NaN     NaN   
3377  8/7/2002       NaN  bonrip   4b  121

In [15]:
hares.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3380 entries, 0 to 3379
Data columns (total 14 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        3380 non-null   object 
 1   time        264 non-null    object 
 2   grid        3380 non-null   object 
 3   trap        3368 non-null   object 
 4   l_ear       3332 non-null   object 
 5   r_ear       3211 non-null   object 
 6   sex         3028 non-null   object 
 7   age         1269 non-null   object 
 8   weight      2845 non-null   float64
 9   hindft      1633 non-null   float64
 10  notes       243 non-null    object 
 11  b_key       3333 non-null   float64
 12  session_id  3380 non-null   int64  
 13  study       3217 non-null   object 
dtypes: float64(3), int64(1), object(10)
memory usage: 369.8+ KB


# Preliminary Exploration

## What are the dimensions of the dataframe and what are the data types of the columns? Do the data types match what you would expect from each column?

In [16]:
print(hares.shape, end = "\n")

(3380, 14)


The dimensions are 3380 observations and 14 variables

## Are there any columns that have a significant number of NA values?

In [17]:
hares.isna().sum()

date             0
time          3116
grid             0
trap            12
l_ear           48
r_ear          169
sex            352
age           2111
weight         535
hindft        1747
notes         3137
b_key           47
session_id       0
study          163
dtype: int64

Counting NAs, we see that a significant amount of NA values in column time, age, hindft, sex, and notes columns.

## What are the minimum and maximum values for  the weight and hind feet measurements?

In [18]:
print('The min and max of weight are \n')
print(hares['weight'].min())
print(hares['weight'].max(),'\n')
print('The min and max of hind feet are \n')
print(hares['hindft'].min())
print(hares['hindft'].max())

The min and max of weight are 

0.0
2365.0 

The min and max of hind feet are 

60.0
160.0


## What are the unique values for some of the categorical columns?

In [29]:
print(hares['grid'].unique())
print(hares['sex'].unique())

['bonrip' 'bonbs' 'Bonbs' 'bonmat' 'bonmat ']
[nan 'M' 'F' '?' 'F?' 'M?' 'pf' 'm' 'f' 'f?' 'm?' 'f ' 'm ']


In [20]:
print(hares['grid'].nunique())
print(hares['sex'].nunique())
print(hares['weight'].nunique())
print(hares['trap'].nunique())

5
12
266
121


- An explroatory question about the data frame you come up with!

### What trap captured the most hares?

# 4) Detecting Messy Values

### a) In the metadata section of the EDI repository, find which are the allowed values for the hares’ sex. Create a small table in a markdown cell showing the values and their definitions.

| Code | Definition |
|------|------------|
| F    | Female     |
| M    | Male       |
| m?    | male not confirmed     |

### b) Get the number of times each unique sex non-NA value appears.

In [28]:
hares['sex'].value_counts(dropna = False)

sex
F      1161
M       730
f       556
m       515
NaN     352
?        40
F?       10
f         4
m         4
f?        3
M?        2
m?        2
pf        1
Name: count, dtype: int64

### c) Check the documentation of value_counts(). What is the purpose of the dropna parameter and what is its default value? Repeat step (a), this time adding the dropna=False parameter to value_counts().

Purpse of the dropna parameter: whether to include counts of rows that contain NA values. Default is True

## d. Discuss with your team the output of the unique value counts. In particular:
- Do the values in the sex column correspond to the values declared in the metadata?
Some values in the `sex` column do not correspond tothe values declared in the metadata.
- What could have been potential causes for multiple codes?
Because different observers entered the observations, there's no limits to the options of recording sex for hares.
- Are there seemingly repated values? If so, what could be the cause?
Capitalization, whitespace, adding question marks could be the cause of repeated values.

### Write code to confirm your suspicions about c-iii.

In [30]:
print(hares['sex'].unique())

[nan 'M' 'F' '?' 'F?' 'M?' 'pf' 'm' 'f' 'f?' 'm?' 'f ' 'm ']


# 5) Brainstorm 

Individually, write step-by-step instructions on how you would wrangle the hares data frame to clean the values in the sex column to have only two classes female and male. Which codes would you assign to each new class? Remember: It’s ok if you don’t know how to code each step - it’s more important to have an idea of what you would like to do.

Discuss your step-by-step instructions with your team.

1)  Inspect the columns
2) Strip whitespace
3) convert all values to lowercase as per the metadata mapping
4) Assume all characters that match the metadata definitions are correct and change any unexpected code to unknown
5) Create a new column with np.select() or np.where()
6) Check results with df.values_counts(dropna = False)

# 6) Clean Values

In [31]:
hares['sex'].unique()

array([nan, 'M', 'F', '?', 'F?', 'M?', 'pf', 'm', 'f', 'f?', 'm?', 'f ',
       'm '], dtype=object)

In [38]:
x = hares.sex

conditions = [
    x.isin(['M','M?','m','m','m ','m_']),
    x.isin(['F','f?','f','F?','f ', 'f_'])]

output_list = ['male','female']

new_column = np.select(conditions, output_list, default = np.nan)


In [41]:
hares['sex_simple'] = new_column

In [42]:
hares['sex_simple'].value_counts()

sex_simple
female    1734
male      1251
nan        395
Name: count, dtype: int64

In [43]:
hares['sex_simple'].value_counts()

sex_simple
female    1734
male      1251
nan        395
Name: count, dtype: int64

# 7) Calcuate Mean Weight

In [ ]:
hares.groupby(by = 'sex_simple')